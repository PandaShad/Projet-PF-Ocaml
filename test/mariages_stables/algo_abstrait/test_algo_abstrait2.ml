open Mariages_stables.Definitions
open Mariages_stables.Algo_abstrait
open Mariages_stables_exemple.Exemple

let () = Random.init 100

module Pioche_aleatoire : PIOCHE = struct
  type 'a t = ('a * bool) array
  let of_list l = 
    let arr = Array.of_list l in
    Array.sort compare arr;
    Array.map (fun x -> (x, true)) arr

  let pioche p =
    let deb = Random.int (Array.length p) in
    let rec essaie i inclus =
      if i=deb && not inclus then None
      else if snd p.(i) then begin
        p.(i) <- (fst p.(i), false) ;
        Format.printf "pioche %s\n" (string_of_homme i);
        Some (fst p.(i))
      end else essaie ((i+1) mod (Array.length p)) false 
    in essaie deb true
  let defausse x p = 
    for i=0 to Array.length p - 1 do
      if fst p.(i) = x then begin
        p.(i) <- (x, true);
        Format.printf "defausse %s\n" (string_of_homme i)
      end
    done
end

module M = Algo(Pioche_aleatoire)
module M2 = Algo(Pile)
module M3 = Algo(File)

let rend_verbeux f prefere_f h h' =
  Format.printf "demande Ã  %s de choisir entre %s et %s\n" 
  (string_of_femme f) (string_of_homme (min h h')) (string_of_homme (max h h'));
  prefere_f h h'
let algo ?(affiche_config=false) entree = 
  ignore affiche_config;
  print_entree entree; 
  let e = { entree with prefere= Array.mapi rend_verbeux entree.prefere} in
  let sortie = M.run e in
  assert (sortie = M2.run entree);
  assert (sortie = M3.run entree);
  print_sortie sortie

let () = algo exemple_knuth