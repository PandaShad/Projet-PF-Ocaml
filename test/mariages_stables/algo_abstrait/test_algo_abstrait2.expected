ENTREE {
  n=4
  A : c,b,d,a
  B : b,a,c,d
  C : b,d,a,c
  D : c,a,d,b
  a : A,B,D,C
  b : C,A,D,B
  c : C,B,D,A
  d : B,A,C,D
}
pioche A
pioche B
pioche D
demande à c de choisir entre A et D
defausse A
pioche C
demande à b de choisir entre B et C
defausse B
pioche A
demande à b de choisir entre A et C
defausse A
pioche A
pioche B
SORTIE { Ad Ba Cb Dc }
